hazelcast-sql/src/main/java/com/hazelcast/sql/impl/expression/ExpressionEvalContextImpl.java

```diff
@@ -50,6 +55,15 @@ public ExpressionEvalContextImpl(
         this.nodeEngine = requireNonNull(nodeEngine);
     }
 
+    public ExpressionEvalContextImpl(
```

Comment:
constructor without ctx/ssc creates insecure expression evaluation context. It should be removed, and places where it is used fixed

Reply:
see https://github.com/hazelcast/hazelcast-enterprise/pull/6520/commits/285e52eb5038d4ef01c23127c3fccabfe1d96f87

Reply:
you should not pass constant `NoOpSqlSecurityContext.INSTANCE` to the constructor because it has the same issue: you create insecure context. So you did not solve the original problem ("constructor without ctx/ssc creates insecure expression evaluation context") only moved it somewhere else.

Reply:
this is still wrong: https://github.com/hazelcast/hazelcast-enterprise/pull/6520/commits/210f7fa2450db1fac9ef220c4f9703947c246fa0

I find security issues only by analysis misuse of bad API.

Reply:
Cool

Reply:
> Cool

I'd prefer to have API that does not allow for misuse (eg. does not have methods to create objects that are insecure). Did you check if that is possible?

Reply:
@k-jamroz all sources accept `ssc` directed from query request, only tests have insecure ctor calls

Reply:
except join and update by key. You fixed that usages but the API (EECI constructor) still allows passing `NoOpSqlSecurityContext.INSTANCE` which is tempting when you do not have `Subject` at hand. And that was exactly what caused problems with updates and joins.

Reply:
Sorry, but right now I don't understand what you want to change... 
Yes, `NoOpSqlSecurityContext.INSTANCE` is still possible to be passed.

Reply:
> Sorry, but right now I don't understand what you want to change... Yes, `NoOpSqlSecurityContext.INSTANCE` is still possible to be passed.

https://github.com/hazelcast/hazelcast/pull/25477#discussion_r1339889236

Reply:
WDYT about https://github.com/hazelcast/hazelcast/pull/25477/commits/1a8d70368b46cee1cb107648f17b11de8be1888d?

Reply:
3-arg `ExpressionEvalContextImpl` still creates insecure object and nothing hints to the API users (i.e. us) that something else has to be considered, so you returned to the initial insecure-by-default state

Reply:
did you consider making EEC serializable (IDS)? Is there any problem with that?

Reply:
> did you consider making EEC serializable (IDS)? Is there any problem with that?

No, because there is nothing to serialize, except `arguments`

Reply:
subject?

Reply:
And subject

Reply:
There is no subject

Reply:
there is. you cannot deserialize EEC to be secure without having a subject. And `subject` is always passed alongside other data needed to create EEC (like `arguments`). But instead of serializing EEC as a whole we serialize `argument` and `subject` independently in every class that needs EEC serialization (breaking encapsulation and DRY by the way) and then do some hackery in `setSerializationService`